File: Mcma.Api/McmaApiController.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Routing;
using Newtonsoft.Json.Linq;
using Mcma.Core.Serialization;
using Mcma.Core.Logging;

namespace Mcma.Api
{
    public class McmaApiController<TRequest> where TRequest : McmaApiRequest
    {
        private List<McmaApiRoute<TRequest>> Routes { get; } = new List<McmaApiRoute<TRequest>>();

        private IDictionary<string, string> GetDefaultResponseHeaders()
            => new Dictionary<string, string>
            {
                ["Date"] = DateTime.UtcNow.ToString(),
                ["Content-Type"] = "application/json",
                ["Access-Control-Allow-Origin"] = "*"
            };

        public void AddRoute(string method, string path, Func<TRequest, McmaApiResponse, Task> handler)
        {
            Routes.Add(new McmaApiRoute<TRequest>(method, path, handler));
        }

        public async Task<McmaApiResponse> HandleRequestAsync(TRequest request)
        {
            var response = new McmaApiResponse
            {
                StatusCode = (int)HttpStatusCode.OK,
                StatusMessage = null,
                Headers = GetDefaultResponseHeaders(),
                JsonBody = null
            };

            var pathMatched = false;
            var methodMatched = false;

            try
            {
                var requestBodyOk = true;

                if (!string.IsNullOrEmpty(request.Body))
                {
                    try
                    {
                        request.JsonBody = JObject.Parse(request.Body);
                    }
                    catch (Exception ex)
                    {
                        response.StatusCode = (int)HttpStatusCode.BadRequest;
                        response.JsonBody = new McmaApiError(response.StatusCode, ex.ToString(), request.Path).ToMcmaJson();
                        requestBodyOk = false;
                    }
                }

                if (requestBodyOk)
                {
                    var methodsAllowed = string.Empty;

                    foreach (var route in Routes)
                    {
                        var pathVariables = new RouteValueDictionary();
                        if (route.Template.TryMatch(request.Path, pathVariables))
                        {
                            pathMatched = true;

                            if (methodsAllowed.Length > 0)
                                methodsAllowed += ", ";
                            methodsAllowed += request.HttpMethod;

                            if (route.HttpMethod == request.HttpMethod)
                            {
                                methodMatched = true;

                                request.PathVariables = pathVariables;

                                await route.Handler(request, response);
                                break;
                            }
                        }
                    }

                    if (!pathMatched)
                    {
                        response.StatusCode = (int)HttpStatusCode.NotFound;
                        response.Headers = GetDefaultResponseHeaders();
                        response.JsonBody = new McmaApiError(response.StatusCode, "Resource not found on path '" + request.Path + "'.", request.Path).ToMcmaJson();
                    }
                    else if (!methodMatched)
                    {
                        if (!methodsAllowed.Contains("OPTIONS"))
                        {
                            if (methodsAllowed.Length > 0)
                                methodsAllowed += ", ";
                            methodsAllowed += "OPTIONS";
                        }

                        if (request.HttpMethod == "OPTIONS")
                        {
                            response.StatusCode = (int)HttpStatusCode.OK;
                            response.Headers = GetDefaultResponseHeaders();

                            string corsMethod = null;
                            string corsHeaders = null;

                            foreach (var prop in request.Headers.Keys)
                            {
                                if (prop.ToLower() == "access-control-request-method")
                                    corsMethod = request.Headers[prop];
                                if (prop.ToLower() == "access-control-request-headers")
                                    corsHeaders = request.Headers[prop];
                            }

                            if (corsMethod != null)
                            {
                                response.Headers["Access-Control-Allow-Methods"] = methodsAllowed;

                                if (corsHeaders != null)
                                    response.Headers["Access-Control-Allow-Headers"] = corsHeaders;
                            }
                            else
                                response.Headers["Allow"] = methodsAllowed;
                        }
                        else
                        {
                            response.StatusCode = (int)HttpStatusCode.MethodNotAllowed;
                            response.Headers = GetDefaultResponseHeaders();
                            response.Headers["Allow"] = methodsAllowed;
                            response.JsonBody = new McmaApiError(response.StatusCode, "Method '" + request.HttpMethod + "' not allowed on path '" + request.Path, request.Path).ToMcmaJson();
                        }
                    }
                    else if ((response.StatusCode / 200 << 0) * 200 == 400)
                    {
                        response.Headers = GetDefaultResponseHeaders();
                        response.JsonBody = new McmaApiError(response.StatusCode, response.StatusMessage, request.Path).ToMcmaJson();
                    }
                }
            }
            catch (Exception ex)
            {
                Logger.Error(ex.ToString());

                response.StatusCode = (int)HttpStatusCode.InternalServerError;
                response.Headers = GetDefaultResponseHeaders();
                response.JsonBody = new McmaApiError(response.StatusCode, ex.ToString(), request.Path).ToMcmaJson();
            }

            if (response.JsonBody != null)
                response.Body = response.JsonBody.ToString();

            return response;
        }
    }
}

File: nodejs/mcma-api/lib/mcma-api-controller.js
//"use strict";
const { HttpStatusCode, getStatusError } = require("./http-statuses");

class McmaApiError {
    constructor(statusCode, message, path) {
        this["@type"] = "ApiError";
        this.timestamp = new Date().toISOString();
        this.status = statusCode;
        this.error = getStatusError(statusCode);
        this.message = message;
        this.path = path;
    }
}

const getDefaultResponseHeaders = () => {
    return {
        "Date": new Date().toUTCString(),
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*"
    }
}

class McmaApiController {
    constructor(routes) {
        this.routes = routes || [];
    }

    async handleRequest(requestContext) {
        const request = requestContext.request;
        const response = requestContext.response;

        response.headers = getDefaultResponseHeaders();

        let pathMatched = false;
        let methodMatched = false;

        try {
            let requestBodyOK = true;

            if (request.body) {
                try {
                    request.body = JSON.parse(request.body);
                } catch (error) {
                    response.statusCode = HttpStatusCode.BAD_REQUEST;
                    response.body = new McmaApiError(response.statusCode, error.message, request.path);
                    requestBodyOK = false;
                }
            }

            if (requestBodyOK) {
                let methodsAllowed = "";

                for (const route of this.routes) {
                    if (route.template.test(request.path, { strict: true })) {
                        pathMatched = true;

                        if (methodsAllowed) {
                            methodsAllowed += ", ";
                        }
                        methodsAllowed += route.httpMethod;

                        if (route.httpMethod === request.httpMethod) {
                            methodMatched = true;

                            request.pathVariables = route.template.fromUri(request.path, { strict: true });

                            await route.handler(requestContext);
                            break;
                        }
                    }
                }

                if (!pathMatched) {
                    response.statusCode = HttpStatusCode.NOT_FOUND;
                    response.headers = getDefaultResponseHeaders();
                    response.body = new McmaApiError(response.statusCode, "Resource not found on path '" + request.path + "'.", request.path);
                } else if (!methodMatched) {
                    if (!methodsAllowed.includes("OPTIONS")) {
                        if (methodsAllowed) {
                            methodsAllowed += ", ";
                        }
                        methodsAllowed += "OPTIONS";
                    }

                    if (request.httpMethod === "OPTIONS") {
                        response.statusCode = HttpStatusCode.OK;
                        response.headers = getDefaultResponseHeaders();

                        let corsMethod;
                        let corsHeaders;

                        // checking if its a CORS pre-flight request
                        for (const prop in request.headers) {
                            if (prop.toLowerCase() === "access-control-request-method") {
                                corsMethod = request.headers[prop];
                            }
                            if (prop.toLowerCase() === "access-control-request-headers") {
                                corsHeaders = request.headers[prop];
                            }
                        }

                        if (corsMethod) { // handling CORS pre-flight request
                            response.headers["Access-Control-Allow-Methods"] = methodsAllowed;

                            if (corsHeaders) {
                                response.headers["Access-Control-Allow-Headers"] = corsHeaders;
                            }
                        } else { // handling regular OPTIONS request
                            response.headers["Allow"] = methodsAllowed;
                        }
                    } else {
                        response.statusCode = HttpStatusCode.METHOD_NOT_ALLOWED;
                        response.headers = getDefaultResponseHeaders();
                        response.headers["Allow"] = methodsAllowed;
                        response.body = new McmaApiError(response.statusCode, "Method '" + request.httpMethod + "' not allowed on path '" + request.path + "'.", request.path);
                    }
                } else if ((response.statusCode / 200 << 0) * 200 === 400) {
                    response.headers = getDefaultResponseHeaders();
                    response.body = new McmaApiError(response.statusCode, response.statusMessage, request.path);
                } else {
                    response.statusCode = HttpStatusCode.OK;
                }
            }
        } catch (error) {
            console.error(error);

            response.statusCode = HttpStatusCode.INTERNAL_ERROR;
            response.headers = getDefaultResponseHeaders();
            response.body = new McmaApiError(response.statusCode, error.message, request.path);
        }

        if (response.body) {
            response.body = JSON.stringify(response.body);
        }
    }
}

module.exports = {
    McmaApiController
};
